lib/LANraragi/Controller/Api/Archive.pm

```diff
@@ -107,6 +114,173 @@ sub serve_file {
     $self->render_file( filepath => $file );
 }
 
+# Create a file archive along with any metadata.
+# adapted from Upload.pm
+sub create_archive {
+    my $self = shift;
+
+    my $logger = get_logger( "Archive API ", "lanraragi");
+    my $redis   = LANraragi::Model::Config->get_redis;
+
+    # receive uploaded file
+    my $upload              = $self->req->upload('file');
+    my $expected_checksum   = $self->req->param('file_checksum'); # optional
+
+    # require file
+    if ( ! defined $upload || !$upload ) {
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "No file attached"
+            },
+            status => 400
+        );
+    }
+
+    # checksum verification stage.
+    if ( $expected_checksum ) {
+        my $file_content        = $upload->slurp;
+        my $actual_checksum     = sha1_hex($file_content);
+        if ( $expected_checksum ne $actual_checksum ) {
+            return $self->render(
+                json => {
+                    operation   => "upload",
+                    success     => 0,
+                    error       => "Checksum mismatch: expected $expected_checksum, got $actual_checksum."
+                },
+                status => 417
+            );
+        }
+    }
+
+    my $filename        = $upload->filename;
+    my $uploadMime      = $upload->headers->content_type;
+
+    # utf downgrade (see LANraragi::Utils::Minion)
+    $filename = encode('UTF-8', $filename);
```

Comment:
Why is this necessary? To me it looks like you're just passing the filename through utf-8 and back out for no real reason now.

Reply:
tbh I don't think I can confidently explain this part of the code because I don't get perl encoding, it was a product of experimentation and was necessary because it prevented errors and ensured that API uploads behave the same as manual uploads (i.e. `LANraragi::Controller::Upload::process_upload`). I'll describe a timeline explaining the rationale.

For the minimal example consider an archive with a special character called "～.cbz". Note that this type of file can already be uploaded successfully through the manual upload portal. (It goes without saying that the behavior of the manual upload dictates the expected behavior of the API upload.)

The entire utf encode/downgrade block exists to guard against issues with redis when a file with a special character like this is uploaded. Redis locking is required before any writes to guard against race conditions. Without the utf block, an unsanitized filename with special characters will cause redis to throw an exception.

When utf-downgrade is (re)introduced, any upload that would've incurred a redis exception is now "handled" at the downgrade stage, which returns an HTTP error. All filenames that pass this downgrade test also do not trigger redis errors. However, this does not *fix* the issue that a file with special characters *should not* fail a downgrade in the API at all, when it never failed this stage during a manual upload, so why does this happen

During the manual process, the file is written to a temporary location first, and the filename is extracted using perl's built-in `File::Find` module. This means somewhere perl already performed encoding/normalization work on the file, which is not present in the API, hence it passing the downgrade stage. (Alternatively, this encoding may be happening during the javascript phase of the upload.)

> to verify my assertion that `File::Find` performs encoding work (which I haven't done since I think the issue was resolved), you can do this: in the API stage, take the file bytes and filename and write them to a temporary directory and read it again with `File::Find` overwriting the filename. If I'm correct then this should pass utf downgrade.

The obvious solution would be to just copy the `File::Find` logic of the manual upload to the API, but the API method cannot afford to move the file to disk due to Redis, as it needs to happen before any writes. So a way to do this is to emulate the effect of `File::Find` and manually encode the file, hence the utf encode stage. Adding this stage makes the API have expected behavior.

Reply:
right, the File::Find way was designed to make sure we get the accurate filesystem representation of the file, since filesystems can and will mangle names.  

But here, you can't wait for that since you need to lock the database on the filename... If the encode/downgrade works I don't mind it I suppose, I don't want to hang the PR forever over this.  

Although... have you tried just using `redis_encode`/`redis_decode`? Those are the intended methods for treating strings before putting them in/taking them out of the database, the downgrade is very much a nuclear/unsafe option.

Reply:
> Although... have you tried just using `redis_encode`/`redis_decode`? Those are the intended methods for treating strings before putting them in/taking them out of the database, the downgrade is very much a nuclear/unsafe option.

Ah... their existence kind of just left my mind 😅 replaced them and tested it, seems like there's no issue. Unfortunately/admittedly, I'm not knowledgable enough about encodings

Reply:
Thanks for taking the time to look over all this, sorry if the PR has gotten too long.

Reply:
No problem at all - Thanks for having put up with my review for all this time.